src/transport/web-transport.ts

```diff
@@ -0,0 +1,389 @@
+            if (status >= 400) throw PubNubAPIError.create(transportResponse);
+
+            this.logRequestProcessProgress(request, undefined, new Date().getTime() - start, responseBody);
+
+            return transportResponse;
+          })
+          .catch((error) => {
+            throw PubNubAPIError.create(error);
+          });
+      }),
+      cancellation,
+    ];
+  }
+
+  request(req: TransportRequest): TransportRequest {
+    return req;
+  }
+
+  /**
+   * Send network request using preferred API.
+   *
+   * @param req - Transport API agnostic request object with prepared body and URL.
+   * @param controller - Request cancellation controller (for long-poll requests).
+   *
+   * @returns Promise which will be resolved or rejected at the end of network request processing.
+   *
+   * @internal
+   */
+  private async sendRequest(req: WebTransportRequest, controller: WebCancellationController): Promise<Response> {
+    if (this.transport === 'fetch') return this.sendFetchRequest(req, controller);
+    return this.sendXHRRequest(req, controller);
+  }
+
+  /**
+   * Send network request using legacy XMLHttpRequest API.
+   *
+   * @param req - Transport API agnostic request object with prepared body and URL.
+   * @param controller - Request cancellation controller (for long-poll requests).
+   *
+   * @returns Promise which will be resolved or rejected at the end of network request processing.
+   *
+   * @internal
+   */
+  private async sendFetchRequest(req: WebTransportRequest, controller: WebCancellationController): Promise<Response> {
+    let timeoutId: ReturnType<typeof setTimeout> | undefined;
+
+    const requestTimeout = new Promise<Response>((_, reject) => {
+      timeoutId = setTimeout(() => {
+        clearTimeout(timeoutId);
+
+        reject(new Error('Request timeout'));
+        controller.abort();
+      }, req.timeout * 1000);
+    });
+
+    const request = new Request(req.url, {
+      method: req.method,
+      headers: req.headers,
+      redirect: 'follow',
+      body: req.body,
+    });
+
+    return Promise.race([
+      WebTransport.originalFetch(request, {
+        signal: controller.abortController.signal,
+        credentials: 'omit',
+        cache: 'no-cache',
+      }).then((response) => {
+        if (timeoutId) clearTimeout(timeoutId);
+        return response;
+      }),
+      requestTimeout,
+    ]);
+  }
+
+  /**
+   * Send network request using legacy XMLHttpRequest API.
+   *
+   * @param req - Transport API agnostic request object with prepared body and URL.
+   * @param controller - Request cancellation controller (for long-poll requests).
+   *
+   * @returns Promise which will be resolved or rejected at the end of network request processing.
+   *
+   * @internal
+   */
+  private async sendXHRRequest(req: WebTransportRequest, controller: WebCancellationController): Promise<Response> {
+    return new Promise<Response>((resolve, reject) => {
+      const xhr = new XMLHttpRequest();
```

Comment:
Here we go again!!!! :(

Reply:
It is a backup for really messed up case with APM tools and some weird browser restrictions which maybe enforced... `fetch` is shiny and better :)